### PR TITLE
fix(tui): list saved models from all providers in /model picker

### LINKED ISSUE #2596: TUI /model picker does not show custom models from other providers in config.toml
## Problem

When a user configures a custom model in `config.toml` under a different provider (e.g. `[models."moonshot-ai/kimi-k2.6"]` with `provider = "moonshot"` while the active provider is `deepseek`), the TUI `/model` picker only shows models from the currently active provider. The custom model is invisible in the picker even though:

- `codewhale model list` lists it correctly as `kimi-k2.6 (moonshot)`
- `codewhale model resolve kimi-k2.6` resolves it without error
- `/model kimi-k2.6` typed directly in the TUI switches to it successfully
- The model works when used

## Root cause

In `crates/tui/src/tui/model_picker.rs`, the picker gets its model list via:

```rust
let model_ids = picker_model_ids_for_provider(app.api_provider);
```

This returns a static, hardcoded list of model IDs for *only* the current provider. Custom models registered under `[models."..."]` in `config.toml` are never merged into this list.

The fallback path (`show_custom_model_row`) only activates for the *currently selected* model if it is not found in the static list — so a custom model that has never been selected is invisible in the picker entirely.

By contrast, the CLI `codewhale model list` aggregates all configured models across all providers, and the slash command `/model <id>` resolves model IDs globally.

## Expected behavior

The `/model` picker should surface all configured models, including custom models from `config.toml` regardless of the active provider. Options:

**Option A (minimal):** Read `config.toml` `[models.*]` entries and append any custom models that are not already in the static provider list to the picker model list.

**Option B (broader):** Merge models from all configured providers into the picker, grouping by provider or showing the provider name alongside the model name (as `codewhale model list` does with `kimi-k2.6 (moonshot)`).

**Option C (hybrid):** Keep the picker scoped to the current provider for the main list, but add a "custom models" section at the bottom showing all entries from `[models.*]` in config.

## Workaround

- `/provider moonshot` then `/model` — the picker will show the model once on the correct provider
- `/model kimi-k2.6` typed directly resolves across providers

## Related

- #2088 Cross-provider model catalog (`codewhale models --all`) — adjacent CLI feature, different surface
- #2026 Model picker: whale-size names — the picker itself, but not about cross-provider visibility

## Environment

- codewhale 0.8.50
- Custom model entry: `[models."moonshot-ai/kimi-k2.6"]` with `provider = "moonshot"`, active provider `deepseek`

### PR DESCRIPTION
**Pull request by ousamabenyounes**
## Summary

- The `/model` picker only listed the active provider's models plus that provider's saved model, so a custom model saved under a *different* provider in config (e.g. `kimi-k2.6` under `moonshot` while the active provider is `deepseek`) was invisible in the picker — even though `/model kimi-k2.6` typed directly resolves and the model runs.
- `picker_model_rows_for_app` now appends cross-provider saved models from `provider_models` as a clearly labelled tail (`<Provider> saved`) after the active provider's list. The active provider's models still come first.
- Selecting a cross-provider row switches CodeWhale to that provider on apply, reusing the existing `resolved_provider` / `build_event` path (no new apply logic). Config keys that map to no known provider are skipped since they cannot be applied.

Fixes #2596.

## Design note

`picker_hides_saved_models_from_other_providers` previously asserted these models stay hidden. That test is repurposed into `picker_lists_saved_models_from_other_providers` to reflect the issue's requested behaviour (Option C — keep the active list first, add a labelled cross-provider tail). The apply path already supported a row carrying its own provider, so this only changes what rows are built, not how they apply.

## Test verification (RED -> GREEN)

New/updated test: `picker_lists_saved_models_from_other_providers`.

RED (prod fix reverted, cross-provider loop removed):
```
test tui::model_picker::tests::picker_lists_saved_models_from_other_providers ... FAILED
panicked at crates/tui/src/tui/model_picker.rs: assert model_ids.contains("deepseek-v4-pro")
test result: FAILED. 28 passed; 1 failed
```

GREEN (fix applied):
```
test tui::model_picker::tests::picker_lists_saved_models_from_other_providers ... ok
test tui::model_picker::tests::picker_skips_unknown_provider_saved_models ... ok
test result: ok. 29 passed; 0 failed
```

## Testing

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --workspace --all-targets --all-features`
- [x] `cargo test --workspace --all-features` — 4608 passed, 0 failed (baseline 4607; +1 new test, no regressions)

## Checklist

- [x] Updated docs or comments as needed
- [x] Added or updated tests where relevant
- [x] Verified TUI behavior manually if UI changes

<details open><summary><h3>Greptile Summary</h3></summary>

Makes cross-provider saved models visible in the `/model` picker by appending them as a labelled tail after the active provider's list, and switches the provider on apply via the existing `resolved_provider` / `build_event` path.

- `picker_model_rows_for_app` now iterates `provider_models`, parses each key with `ApiProvider::parse`, skips the active provider and unknown keys, sorts the remainder deterministically by config key, and pushes one labelled row per entry.
- `picker_hides_saved_models_from_other_providers` is repurposed into `picker_lists_saved_models_from_other_providers` with assertions for ordering, row-level provider attribution, and provider-switch on apply; a new `picker_skips_unknown_provider_saved_models` test covers the filter-out path.
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge; the apply path is unchanged and all existing tests pass with one new test converted from failing to green.

The change is contained to row-building logic with no modification to how events are emitted or applied. There is one untested edge case: when the user's active model was set via a direct /model command to a cross-provider ID rather than through the picker, that model now appears both as a labelled tail row and as the current (custom) row — a cosmetic duplication that does not cause incorrect behaviour but could confuse users.

crates/tui/src/tui/model_picker.rs — specifically the ModelPickerView::new selection logic and the interaction between show_custom_model_row and the new cross-provider tail rows.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| crates/tui/src/tui/model_picker.rs | Adds a cross-provider saved-model tail to the picker row list, sorts it deterministically by config key, filters out unknown providers, and updates tests to assert the new behaviour. The core fix is correct; a minor edge case exists where the model can appear twice if set via direct command rather than the picker. |

</details>

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[picker_model_rows_for_app] --> B[Push auto row]
    B --> C[Push active provider known models]
    C --> D{Active provider saved model exists?}
    D -- Yes --> E[Push Active-Provider saved row]
    D -- No --> F
    E --> F[Iterate all provider_models entries]
    F --> G{ApiProvider::parse succeeds?}
    G -- No unknown key --> H[Skip entry]
    G -- Yes --> I{Same as active provider?}
    I -- Yes --> H
    I -- No --> J[Collect cross-provider entry]
    J --> K[Sort by config key string]
    K --> L{model string empty after trim?}
    L -- Yes --> H
    L -- No --> M[Push Provider-saved row with provider set]
    M --> N[Return rows]
    H --> N
```
</details>

<a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22hmbown%2Fcodewhale%22%20on%20the%20existing%20branch%20%22fix%2Fissue-2596%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22fix%2Fissue-2596%22.%0A%0AFix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Acrates%2Ftui%2Fsrc%2Ftui%2Fmodel_picker.rs%3A71-78%0A**Cross-provider%20model%20shown%20twice%20when%20set%20via%20direct%20command**%0A%0AThe%20%60selected_model_idx%60%20search%20%28line%2071%E2%80%9374%29%20matches%20only%20rows%20whose%20%60provider%60%20is%20%60None%60%20%28auto%29%20or%20equal%20to%20%60app.api_provider%60.%20Cross-provider%20rows%20added%20by%20the%20new%20tail%20loop%20never%20satisfy%20%60row.provider%20%3D%3D%20Some%28app.api_provider%29%60%2C%20so%20when%20%60app.model%60%20was%20set%20to%20a%20cross-provider%20model%20through%20a%20direct%20%60%2Fmodel%20%3Cid%3E%60%20command%20%28not%20via%20the%20picker%29%2C%20%60selected_model_idx%60%20stays%20%60None%60%2C%20%60show_custom_model_row%60%20becomes%20%60true%60%2C%20and%20the%20model%20ends%20up%20appearing%20twice%3A%20once%20as%20a%20labelled%20%22X%20saved%22%20tail%20row%20and%20once%20as%20the%20%22current%20%28custom%29%22%20row%20at%20the%20bottom.%20The%20PR%20description%20specifically%20mentions%20this%20path%20%28%22even%20though%20%60%2Fmodel%20kimi-k2.6%60%20typed%20directly%20resolves%20and%20the%20model%20runs%22%29%2C%20so%20the%20scenario%20is%20realistic.%20The%20fix%20would%20be%20to%20let%20the%20search%20also%20accept%20a%20cross-provider%20match%2C%20or%20to%20suppress%20%60show_custom_model_row%60%20when%20the%20model%20is%20already%20present%20as%20any%20tail%20row.%0A%0ANo%20test%20exercises%20the%20case%20where%20%60app.model%60%20equals%20a%20cross-provider%20saved%20model%20while%20%60api_provider%60%20is%20still%20the%20original%20provider.%0A%0A&repo=hmbown%2Fcodewhale&pr=2869&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Acrates%2Ftui%2Fsrc%2Ftui%2Fmodel_picker.rs%3A71-78%0A**Cross-provider%20model%20shown%20twice%20when%20set%20via%20direct%20command**%0A%0AThe%20%60selected_model_idx%60%20search%20%28line%2071%E2%80%9374%29%20matches%20only%20rows%20whose%20%60provider%60%20is%20%60None%60%20%28auto%29%20or%20equal%20to%20%60app.api_provider%60.%20Cross-provider%20rows%20added%20by%20the%20new%20tail%20loop%20never%20satisfy%20%60row.provider%20%3D%3D%20Some%28app.api_provider%29%60%2C%20so%20when%20%60app.model%60%20was%20set%20to%20a%20cross-provider%20model%20through%20a%20direct%20%60%2Fmodel%20%3Cid%3E%60%20command%20%28not%20via%20the%20picker%29%2C%20%60selected_model_idx%60%20stays%20%60None%60%2C%20%60show_custom_model_row%60%20becomes%20%60true%60%2C%20and%20the%20model%20ends%20up%20appearing%20twice%3A%20once%20as%20a%20labelled%20%22X%20saved%22%20tail%20row%20and%20once%20as%20the%20%22current%20%28custom%29%22%20row%20at%20the%20bottom.%20The%20PR%20description%20specifically%20mentions%20this%20path%20%28%22even%20though%20%60%2Fmodel%20kimi-k2.6%60%20typed%20directly%20resolves%20and%20the%20model%20runs%22%29%2C%20so%20the%20scenario%20is%20realistic.%20The%20fix%20would%20be%20to%20let%20the%20search%20also%20accept%20a%20cross-provider%20match%2C%20or%20to%20suppress%20%60show_custom_model_row%60%20when%20the%20model%20is%20already%20present%20as%20any%20tail%20row.%0A%0ANo%20test%20exercises%20the%20case%20where%20%60app.model%60%20equals%20a%20cross-provider%20saved%20model%20while%20%60api_provider%60%20is%20still%20the%20original%20provider.%0A%0A&repo=hmbown%2Fcodewhale&pr=2869&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Acrates%2Ftui%2Fsrc%2Ftui%2Fmodel_picker.rs%3A71-78%0A**Cross-provider%20model%20shown%20twice%20when%20set%20via%20direct%20command**%0A%0AThe%20%60selected_model_idx%60%20search%20%28line%2071%E2%80%9374%29%20matches%20only%20rows%20whose%20%60provider%60%20is%20%60None%60%20%28auto%29%20or%20equal%20to%20%60app.api_provider%60.%20Cross-provider%20rows%20added%20by%20the%20new%20tail%20loop%20never%20satisfy%20%60row.provider%20%3D%3D%20Some%28app.api_provider%29%60%2C%20so%20when%20%60app.model%60%20was%20set%20to%20a%20cross-provider%20model%20through%20a%20direct%20%60%2Fmodel%20%3Cid%3E%60%20command%20%28not%20via%20the%20picker%29%2C%20%60selected_model_idx%60%20stays%20%60None%60%2C%20%60show_custom_model_row%60%20becomes%20%60true%60%2C%20and%20the%20model%20ends%20up%20appearing%20twice%3A%20once%20as%20a%20labelled%20%22X%20saved%22%20tail%20row%20and%20once%20as%20the%20%22current%20%28custom%29%22%20row%20at%20the%20bottom.%20The%20PR%20description%20specifically%20mentions%20this%20path%20%28%22even%20though%20%60%2Fmodel%20kimi-k2.6%60%20typed%20directly%20resolves%20and%20the%20model%20runs%22%29%2C%20so%20the%20scenario%20is%20realistic.%20The%20fix%20would%20be%20to%20let%20the%20search%20also%20accept%20a%20cross-provider%20match%2C%20or%20to%20suppress%20%60show_custom_model_row%60%20when%20the%20model%20is%20already%20present%20as%20any%20tail%20row.%0A%0ANo%20test%20exercises%20the%20case%20where%20%60app.model%60%20equals%20a%20cross-provider%20saved%20model%20while%20%60api_provider%60%20is%20still%20the%20original%20provider.%0A%0A&pr=2869&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=3">Fix All in Cursor</picture></a>

<sub>Reviews (3): Last reviewed commit: ["fix(tui): use sort\_by\_key to satisfy cli..."](https://github.com/hmbown/codewhale/commit/97f6e0b2e5f9ed12c4e08c63ec655e4753483588) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36039335)</sub>

### CODE REVIEW

File: crates/tui/src/tui/model_picker.rs [hunk 1]
```diff
@@ -385,6 +385,36 @@ fn picker_model_rows_for_app(app: &App) -> Vec<ModelPickerRow> {
...
+        if model.is_empty() {
+            continue;
+        }
+        push_model_row(
+            &mut rows,
+            model.to_string(),
+            Some(provider),
+            format!("{} saved", provider.display_name()),
+        );
+    }
```
[the remark this answers is not in the document]
Comment (ousamabenyounes) [reply crates/tui/src/tui/model_picker.rs]: Done in db0ce5b9 — collapsed the two `ApiProvider::parse` calls into a single `filter_map` that drops unknown keys and the active provider in one pass and carries the parsed provider through to the loop.

One nuance on framing: this is a readability win, not really an optimization. The original code only parsed twice for keys that survive the filter, the map is tiny, and `ApiProvider::parse` is a cheap string match — so there's no measurable perf delta. The value is fewer moving parts and no duplicated parse logic. Behavior is identical: unknown keys are still skipped, the active provider is still excluded, and ordering stays deterministic via the sort on the provider key.

`cargo fmt`/`clippy`/picker tests all green.
